# DAT210x - Programming with Python for DS

## Module6- Lab6

In [106]:
import time
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import sklearn.preprocessing as sp
from sklearn.tree import export_graphviz

### How to Get The Dataset

Grab the DLA HAR dataset from:

- http://groupware.les.inf.puc-rio.br/har
- http://groupware.les.inf.puc-rio.br/static/har/dataset-har-PUC-Rio-ugulino.zip
- A cached copy of the dataset is included in the course repository.

After extracting it out, load up the dataset into dataframe named `X` and do your regular dataframe examination:

In [107]:
X=pd.read_csv('dataset-har-PUC-Rio-ugulino.csv',sep=';')

C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2728: DtypeWarning: Columns (17) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [108]:
X.head()

user gender  age how_tall_in_meters  weight body_mass_index  x1  y1  z1  \
0  debora  Woman   46               1,62      75            28,6  -3  92 -63   
1  debora  Woman   46               1,62      75            28,6  -3  94 -64   
2  debora  Woman   46               1,62      75            28,6  -1  97 -61   
3  debora  Woman   46               1,62      75            28,6  -2  96 -57   
4  debora  Woman   46               1,62      75            28,6  -1  96 -61   

   x2  y2  z2  x3   y3  z3   x4   y4    z4    class  
0 -23  18 -19   5  104 -92 -150 -103  -147  sitting  
1 -21  18 -18 -14  104 -90 -149 -104  -145  sitting  
2 -12  20 -15 -13  104 -90 -151 -104  -144  sitting  
3 -15  21 -16 -13  104 -89 -153 -103  -142  sitting  
4 -13  20 -15 -13  104 -89 -153 -104  -143  sitting

In [109]:
print(X.head())
print("\n")
print(X.shape)
print("\n")
print(X.describe())

     user gender  age how_tall_in_meters  weight body_mass_index  x1  y1  z1  \
0  debora  Woman   46               1,62      75            28,6  -3  92 -63   
1  debora  Woman   46               1,62      75            28,6  -3  94 -64   
2  debora  Woman   46               1,62      75            28,6  -1  97 -61   
3  debora  Woman   46               1,62      75            28,6  -2  96 -57   
4  debora  Woman   46               1,62      75            28,6  -1  96 -61   

   x2  y2  z2  x3   y3  z3   x4   y4    z4    class  
0 -23  18 -19   5  104 -92 -150 -103  -147  sitting  
1 -21  18 -18 -14  104 -90 -149 -104  -145  sitting  
2 -12  20 -15 -13  104 -90 -151 -104  -144  sitting  
3 -15  21 -16 -13  104 -89 -153 -103  -142  sitting  
4 -13  20 -15 -13  104 -89 -153 -104  -143  sitting  


(165633, 19)


                 age         weight             x1             y1  \
count  165633.000000  165633.000000  165633.000000  165633.000000   
mean       38.265146      70.819408     

Encode the gender column such that: `0` is male, and `1` as female:

In [110]:
X.gender=X.gender.map({'Man':1,'Woman':0})
X.gender.unique()

array([0, 1], dtype=int64)

Clean up any columns with commas in them so that they're properly represented as decimals:

Let's take a peek at your data types:

In [111]:
X.dtypes

user                  object
gender                 int64
age                    int64
how_tall_in_meters    object
weight                 int64
body_mass_index       object
x1                     int64
y1                     int64
z1                     int64
x2                     int64
y2                     int64
z2                     int64
x3                     int64
y3                     int64
z3                     int64
x4                     int64
y4                     int64
z4                    object
class                 object
dtype: object

Convert any column that needs to be converted into numeric use `errors='raise'`. This will alert you if something ends up being problematic.

In [112]:
X['how_tall_in_meters']=X['how_tall_in_meters'].str.replace(',', '.')
X['how_tall_in_meters']=pd.to_numeric(X['how_tall_in_meters'],errors='raise')
X['body_mass_index']=X['body_mass_index'].str.replace(',', '.')
X['body_mass_index']=pd.to_numeric(X['body_mass_index'],errors='raise')
print(X.z4.unique())
X=X.drop([122076])
X['z4']=pd.to_numeric(X['z4'],errors='raise')

[-147 -145 -144 -142 -143 -146 -138 -139 -141 -133 -134 -135 -140 -137
 -148 -151 -149 -150 -152 -156 -157 -155 -154 -153 -158 -159 -162 -163
 -161 -160 -164 -165 -136 -126 -125 -122 -132 -167 -170 -169 -175 -173
 -166 -168 -171 -172 -174 -179 -177 -176 -127 -131 -184 -181 -178 -180
 -106 -114 -116 -115 -89 -79 -93 -78 -128 -109 -113 -119 -102 -111 -103
 -130 -129 -182 -121 -186 -183 -187 -190 -188 -189 -185 -192 -191 -193
 -194 -197 -196 -201 -251 -195 -199 -198 -123 -124 -112 -118 -120 -99 -95
 -68 -110 -117 -96 -100 -105 -92 -88 -107 -213 -108 -104 -98 -94 -91 -97
 -69 -101 -86 -82 -66 -77 -90 -56 -74 -83 -81 -219 -200 '-158' '-165'
 '-157' '-160' '-161' '-155' '-162' '-156' '-159' '-163' '-168' '-170'
 '-166' '-172' '-177' '-171' '-169' '-173' '-167' '-174' '-164' '-176'
 '-175' '-154' '-146' '-142' '-137' '-141' '-139' '-148' '-147' '-153'
 '-140' '-138' '-128' '-134' '-150' '-149' '-152' '-151' '-136' '-145'
 '-143' '-182' '-184' '-179' '-186' '-188' '-183' '-191' '-185' '-181'
 

If you find any problematic records, drop them before calling the `to_numeric` methods above.

Okay, now encode your `y` value as a Pandas dummies version of your dataset's `class` column:

In [113]:
y=X[['class']].copy()

In [114]:
y=pd.get_dummies(y,columns=['class'])
y.head()

class_sitting  class_sittingdown  class_standing  class_standingup  \
0              1                  0               0                 0   
1              1                  0               0                 0   
2              1                  0               0                 0   
3              1                  0               0                 0   
4              1                  0               0                 0   

   class_walking  
0              0  
1              0  
2              0  
3              0  
4              0

In fact, get rid of the `user` and `class` columns:

In [115]:
X=X.drop(labels=['user','class'],axis=1)

Let's take a look at your handy-work:

In [116]:
X.describe()

gender            age  how_tall_in_meters         weight  \
count  165632.000000  165632.000000       165632.000000  165632.000000   
mean        0.387956      38.264925            1.639712      70.819431   
std         0.487286      13.183821            0.052820      11.296557   
min         0.000000      28.000000            1.580000      55.000000   
25%         0.000000      28.000000            1.580000      55.000000   
50%         0.000000      31.000000            1.620000      75.000000   
75%         1.000000      46.000000            1.710000      83.000000   
max         1.000000      75.000000            1.710000      83.000000   

       body_mass_index             x1             y1             z1  \
count    165632.000000  165632.000000  165632.000000  165632.000000   
mean         26.188535      -6.649319      88.293591     -93.164449   
std           2.995781      11.616273      23.895881      39.409487   
min          22.000000    -306.000000    -271.000000    -603.000000   
25%          22.000000     -12.000000      78.000000    -120.000000   
50%          28.400000      -6.000000      94.000000     -98.000000   
75%          28.600000       0.000000     101.000000     -64.000000   
max          28.600000     509.000000     533.000000     411.000000   

                  x2             y2             z2             x3  \
count  165632.000000  165632.000000  165632.000000  165632.000000   
mean      -87.827956     -52.065911    -175.055647      17.423517   
std       169.435606     205.160081     192.817111      52.635546   
min      -494.000000    -517.000000    -617.000000    -499.000000   
25%       -35.000000     -29.000000    -141.000000       9.000000   
50%        -9.000000      27.000000    -118.000000      22.000000   
75%         4.000000      86.000000     -29.000000      34.000000   
max       473.000000     295.000000     122.000000     507.000000   

                  y3             z3             x4             y4  \
count  165632.000000  165632.000000  165632.000000  165632.000000   
mean      104.517056     -93.881641    -167.641211     -92.625235   
std        54.155987      45.389770      38.311336      19.968653   
min      -506.000000    -613.000000    -702.000000    -526.000000   
25%        95.000000    -103.000000    -190.000000    -103.000000   
50%       107.000000     -90.000000    -168.000000     -91.000000   
75%       120.000000     -80.000000    -153.000000     -80.000000   
max       517.000000     410.000000     -13.000000      86.000000   

                  z4  
count  165632.000000  
mean     -159.650985  
std        13.221020  
min      -537.000000  
25%      -167.000000  
50%      -160.000000  
75%      -153.000000  
max       -43.000000

You can also easily display which rows have nans in them, if any:

In [117]:
X[pd.isnull(X).any(axis=1)]

Empty DataFrame
Columns: [gender, age, how_tall_in_meters, weight, body_mass_index, x1, y1, z1, x2, y2, z2, x3, y3, z3, x4, y4, z4]
Index: []

Create an RForest classifier named `model` and set `n_estimators=30`, the `max_depth` to 10, `oob_score=True`, and `random_state=0`:

In [118]:
from sklearn.ensemble import RandomForestClassifier
model = RandomForestClassifier(n_estimators=30,max_depth=10, oob_score=True, random_state=0)

Split your data into `test` / `train` sets. Your `test` size can be 30%, with `random_state` 7. Use variable names: `X_train`, `X_test`, `y_train`, and `y_test`:

In [119]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=.3,random_state=7)

### Now the Fun Stuff

In [120]:
print("Fitting...")
s = time.time()

# TODO: train your model on your training set

model.fit(X_train, y_train)


print("Fitting completed in: ", time.time() - s)

Fitting...
Fitting completed in:  12.28407096862793


Output for Male 0 female 1 encoding:
Fitting...
Fitting completed in:  12.515645742416382

Output for Male 1 female 0 encoding:
Fitting...
Fitting completed in:  12.28407096862793

Display the OOB Score of your data:

In [121]:
score = model.oob_score_
print("OOB Score: ", round(score*100, 3))

OOB Score:  98.771


Output for Male 0 female 1 encoding:
OOB Score:  98.744
Output for Male 1 female 0 encoding:
OOB Score:  98.771


In [122]:
print("Scoring...")
s = time.time()

# TODO: score your model on your test set

score=model.score(X_test, y_test)

print("Score: ", round(score*100, 3))
print("Scoring completed in: ", time.time() - s)

Scoring...
Score:  95.939
Scoring completed in:  0.8991992473602295


Output for Male 0 female 1 encoding:
Scoring...
Score:  95.687
Scoring completed in:  1.0583951473236084


Output for Male 1 female 0 encoding:
Scoring...
Score:  95.939
Scoring completed in:  0.8991992473602295

At this point, go ahead and answer the lab questions, then return here to experiment more --

Try playing around with the gender column. For example, encode `gender` `Male:1`, and `Female:0`. Also try encoding it as a Pandas dummies variable and seeing what changes that has. You can also try dropping gender entirely from the dataframe. How does that change the score of the model? This will be a key insight on how your feature encoding alters your overall scoring, and why it's important to choose good ones.

In [123]:
# .. your code changes above ..

After that, try messing with `y`. Right now its encoded with dummies, but try other encoding methods to what effects they have.